File: src/web-ui/src/infrastructure/api/service-api/StartchatAgentAPI.ts
 

import { api } from './ApiClient';
import { createTauriCommandError } from '../errors/TauriCommandError';

 
export interface WorkStateOptions {
   
  analyzeGit?: boolean;
   
  predictNextActions?: boolean;
   
  includeQuickActions?: boolean;
   
  language?: 'Chinese' | 'English';
}

 
export interface GreetingMessage {
   
  title: string;
   
  subtitle: string;
   
  tagline?: string;
}

 
export interface GitWorkState {
   
  currentBranch: string;
   
  unstagedFiles: number;
   
  stagedFiles: number;
   
  unpushedCommits: number;
   
  aheadBehind?: {
    ahead: number;
    behind: number;
  };
   
  modifiedFiles: FileModification[];
}

 
export interface FileModification {
   
  path: string;
   
  changeType: 'Added' | 'Modified' | 'Deleted' | 'Renamed' | 'Untracked';
   
  module?: string;
}

 
export interface WorkItem {
   
  title: string;
   
  description: string;
   
  relatedFiles: string[];
   
  category: 'Backend' | 'Frontend' | 'API' | 'Database' | 'Infrastructure' | 'Testing' | 'Documentation' | 'Other';
   
  icon: string;
}

 
export interface TimeInfo {
   
  minutesSinceLastCommit?: number;
   
  lastCommitTimeDesc?: string;
   
  timeOfDay: 'Morning' | 'Afternoon' | 'Evening' | 'Night';
}

 
export interface CurrentWorkState {
   
  summary: string;
   
  gitState?: GitWorkState;
   
  ongoingWork: WorkItem[];
   
  timeInfo: TimeInfo;
}

 
export interface PredictedAction {
   
  description: string;
   
  priority: 'High' | 'Medium' | 'Low';
   
  icon: string;
   
  isReminder: boolean;
}

 
export interface QuickAction {
   
  title: string;
   
  command: string;
   
  icon: string;
   
  actionType: 'Continue' | 'ViewStatus' | 'Commit' | 'Visualize' | 'Custom';
}

 
export interface WorkStateAnalysis {
   
  greeting: GreetingMessage;
   
  currentState: CurrentWorkState;
   
  predictedActions: PredictedAction[];
   
  quickActions: QuickAction[];
   
  analyzedAt: string;
}

 
export interface WorkStateSummaryResponse {
   
  greetingTitle: string;
   
  currentStateSummary: string;
   
  hasGitChanges: boolean;
   
  unstagedFiles: number;
   
  unpushedCommits: number;
   
  predictedActionsCount: number;
}

 
export class StartchatAgentAPI {
   
  async analyzeWorkState(
    repoPath: string,
    options?: WorkStateOptions
  ): Promise<WorkStateAnalysis> {
    try {
      return await api.invoke('analyze_work_state', {
        request: {
          repoPath,
          options: options || null
        }
      });
    } catch (error) {
      throw createTauriCommandError('analyze_work_state', error, { repoPath, options });
    }
  }

   
  async quickAnalyzeWorkState(repoPath: string): Promise<WorkStateAnalysis> {
    try {
      
      return await api.invoke('quick_analyze_work_state', {
        request: {
          repoPath
        }
      }, {
        retries: 0
      });
    } catch (error) {
      throw createTauriCommandError('quick_analyze_work_state', error, { repoPath });
    }
  }

   
  async generateGreetingOnly(repoPath: string): Promise<WorkStateAnalysis> {
    try {
      return await api.invoke('generate_greeting_only', {
        request: {
          repoPath
        }
      });
    } catch (error) {
      throw createTauriCommandError('generate_greeting_only', error, { repoPath });
    }
  }

   
  async getWorkStateSummary(repoPath: string): Promise<WorkStateSummaryResponse> {
    try {
      return await api.invoke('get_work_state_summary', {
        request: {
          repoPath
        }
      });
    } catch (error) {
      throw createTauriCommandError('get_work_state_summary', error, { repoPath });
    }
  }
}


export const startchatAgentAPI = new StartchatAgentAPI();

File: src/apps/desktop/src/api/startchat_agent_api.rs
//! Startchat Agent API

use log::error;
use tauri::State;
use bitfun_core::function_agents::{
    StartchatFunctionAgent,
    WorkStateAnalysis,
    WorkStateOptions,
};
use serde::{Deserialize, Serialize};
use std::path::Path;

use super::app_state::AppState;

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AnalyzeWorkStateRequest {
    pub repo_path: String,
    pub options: Option<WorkStateOptions>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QuickAnalyzeRequest {
    pub repo_path: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GenerateGreetingRequest {
    pub repo_path: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkStateSummaryResponse {
    pub greeting_title: String,
    pub current_state_summary: String,
    pub has_git_changes: bool,
    pub unstaged_files: u32,
    pub unpushed_commits: u32,
    pub predicted_actions_count: usize,
}

#[tauri::command]
pub async fn analyze_work_state(
    state: State<'_, AppState>,
    request: AnalyzeWorkStateRequest,
) -> Result<WorkStateAnalysis, String> {
    let agent = StartchatFunctionAgent::new(state.ai_client_factory.clone());
    let opts = request.options.unwrap_or_default();
    
    agent
        .analyze_work_state(Path::new(&request.repo_path), opts)
        .await
        .map_err(|e| {
            error!("Work state analysis failed: repo_path={}, error={}", request.repo_path, e);
            e.to_string()
        })
}

#[tauri::command]
pub async fn quick_analyze_work_state(
    state: State<'_, AppState>,
    request: QuickAnalyzeRequest,
) -> Result<WorkStateAnalysis, String> {
    let agent = StartchatFunctionAgent::new(state.ai_client_factory.clone());
    
    agent
        .quick_analyze(Path::new(&request.repo_path))
        .await
        .map_err(|e| {
            error!("Quick work state analysis failed: repo_path={}, error={}", request.repo_path, e);
            e.to_string()
        })
}

#[tauri::command]
pub async fn generate_greeting_only(
    state: State<'_, AppState>,
    request: GenerateGreetingRequest,
) -> Result<WorkStateAnalysis, String> {
    let agent = StartchatFunctionAgent::new(state.ai_client_factory.clone());
    
    agent
        .generate_greeting_only(Path::new(&request.repo_path))
        .await
        .map_err(|e| {
            error!("Generate greeting failed: repo_path={}, error={}", request.repo_path, e);
            e.to_string()
        })
}

#[tauri::command]
pub async fn get_work_state_summary(
    state: State<'_, AppState>,
    request: QuickAnalyzeRequest,
) -> Result<WorkStateSummaryResponse, String> {
    let agent = StartchatFunctionAgent::new(state.ai_client_factory.clone());
    
    let analysis = agent
        .quick_analyze(Path::new(&request.repo_path))
        .await
        .map_err(|e| {
            error!("Failed to get work state summary: repo_path={}, error={}", request.repo_path, e);
            e.to_string()
        })?;
    
    let (unstaged_files, unpushed_commits, has_git_changes) = if let Some(ref git) = analysis.current_state.git_state {
        (
            git.unstaged_files + git.staged_files,
            git.unpushed_commits,
            git.unstaged_files > 0 || git.staged_files > 0 || git.unpushed_commits > 0
        )
    } else {
        (0, 0, false)
    };
    
    Ok(WorkStateSummaryResponse {
        greeting_title: analysis.greeting.title,
        current_state_summary: analysis.current_state.summary,
        has_git_changes,
        unstaged_files,
        unpushed_commits,
        predicted_actions_count: analysis.predicted_actions.len(),
    })
}
